### ISSUE TITLE
eBGP allow own AS (for spines) - plugin pre/post hooks

### ISSUE THREAD

**Issue by jbemmel** [ipspace/netlab#198]
In an eBGP topology that uses the same AS on every spine, I would like each spine to still accept loopback routes to other spines
(for that infamous iBGP peering session...)

There is a knob for that, of course; for example, on SR Linux:
```
   as-path-options:
    allow-own-as: 1
    _annotate_allow-own-as: "Accept loopback route to other spine (same AS)"
```

I can put this in a plugin, but the problem is that it only applies to routes received after this setting. So I'd prefer to include it in the initial BGP config

Any objections to add a node flag "bgp.ebgp_allow_own_as" or something like that?

**Comment by petercrocker** [comment]
I had a network deployed at last role (before I joined) that did this for iBGP evpn and had same AS on all leafs because $vendor. So would be handy.

**Comment by ipspace** [comment]
As they say: "the road to hell (and broken designs) is paved with good intentions".

In this particular case: the reason we're using the same AS number on all spines is to prevent path hunting. Configuring "allow-as-in" or similar destroys that, unless you combine it with a route map that drop anything coming from your own AS unless it's a loopback interface. Hooray, we have job security ;)

If you want to have connectivity between spine switches, use a different AS number on each of them (totally destroying the benefits of using EBGP, but that's a different story).

If someone insists on having IBGP-over-EBGP EVPN design then they should also bite the bullet and NOT have that IBGP session between route reflectors, hopefully having more than two route reflectors in a production network. Unfortunately, as Peter just explained, sanity is a rare commodity in vendor world.

Back to netsim-tools and BGP nerd knobs: while there are good reasons for every nerd knob you introduced (and I'm positive there are people out there who want to build labs with them), they create several challenges:

* Some of those nerd knobs have a half-dozen options (like the **local-as** one) and different vendors implemented different options. We could decide on which set of options to implement and go forward with that, but what happens when someone else wants to have a different set of those options?
* Inevitably, these things won't be implemented on even a majority of the platforms supported by netsim-tools. One of my most frustrating experiences with various open source projects in the last too-many years is inevitably poorly-documented partial implementation (feature X I need works in 30% of the cases, but not where I need them), and I don't want to go down that same path.
* Adding nerd knobs to core modules would result in long and convoluted "feature support" tables and ever-more-complex configuration templates, not to mention higher barrier-to-entry for new device contributions. Or maybe my OCD is kicking in again.

In this particular case, we could go for "bgp-as-hacks" plugin that would implement several AS-related hacks like **local-as**, **allowas-in** and **as-override** (whatever this last one is doing, I know I had to use it a long while ago), and then whoever wants to implement those on a particular platform would be welcome to submit relevant configuration templates.

To summarize this saga: I would prefer to keep core modules as simple as possible and add nerd knobs via optional plugins. Thoughts?

As for "applying routes after receiving them", that's a valid concern. Every time you change BGP policy you usually have to execute "clear ip bgp * soft in" or similar, and there's clearly a need for pre- and post-configuration task hook in deploying custom templates. I could add that in a few days.

**Comment by jbemmel** [comment]
With the new local_as interface level attribute I can build an eBGP session between the spines, voiding the need for allow_own_as

Some people may still want to play around with those knobs, but I agree they have no place in the core. 

Keeping the issue open until we resolve the pre/post hooks for plugins

**Comment by ipspace** [comment]
With 9e8cc16, the **local_as** is automatically copied from node data into interface data, hopefully making your task even simpler.

One of these days I'll add **local_as** logic to neighbor building code within bgp.py -- it doesn't make sense to duplicate the functionality, and it seems to be hard to augment the **bgp.neighbor** list with local_as data as opposed to rebuilding it from scratch.

**Comment by ipspace** [comment]
With 9e8cc16, the **local_as** is automatically copied from node data into interface data, hopefully making your task even simpler.

One of these days I'll add **local_as** logic to neighbor building code within bgp.py -- it doesn't make sense to duplicate the functionality, and it seems to be hard to augment the **bgp.neighbor** list with local_as data as opposed to rebuilding it from scratch.

**Comment by ipspace** [comment]
The "allowas-in" is available in ebgp.utils, the rest of this discussion is now in #363.
